### PR TITLE
fix(Scripts/BlackrockSpire): Urok Doomhowl - improvements:

### LINKED ISSUE #9023: LBRS Urok Doomhowl encounter issues
### What client do you play on?

enUS

### Faction

- [X] Alliance
- [X] Horde

### Content Phase:

- [ ] Generic
- [ ] 1-19
- [ ] 20-29
- [ ] 30-39
- [ ] 40-49
- [X] 50-59

### Current Behaviour

[Urok Doomhowl](https://wowgaming.altervista.org/aowow/?npc=10584), [Urok's Tribute Pile](https://wowgaming.altervista.org/aowow/?object=175621), [Roughshod Pike](https://wowgaming.altervista.org/aowow/?item=12533), [Omokk's Head](https://wowgaming.altervista.org/aowow/?item=12534)

1. The mobs don't attack the Pike with Head
2. The boss is despawning after appear
3. Urok's Tribute Pile with Pike and Head doesn't disapear

### Expected Blizzlike Behaviour

1. The mobs should attack the pike if there is no players around
2. The boss should not despawning after summon
3. Urok's Tribute Pile with Pike and Head should disapear after Urok appear

### Source

Wowhead [Urok Doomhowl](https://www.wowhead.com/npc=10584/urok-doomhowl#comments):
> By Xenomech on 2008/09/10 (Patch 3.0.1)
If you wipe after planting the head on the pike, you don't get a second chance. You'll have to get the head and pike and try all over again.

> By 146568 on 2008/10/26 (Patch 3.0.2)
If you wipe while fighting Urok himself, he'll stay there. So you can wipe safely after the waves of ogres, killing urok later.

> By TwoGlassEyes on 2008/11/01 (Patch 3.0.2)
Also important to note, if you let the spawns take down the pike, they stop spawning and Urok will NOT spawn.

https://youtu.be/2RqoKtMHavk?t=2692
https://youtu.be/L8a-9CYQk1g?t=4165

### Steps to reproduce the problem

.go game 99791
.additem 12533
.additem 12534

### Extra Notes

Related to https://github.com/azerothcore/azerothcore-wotlk/issues/8712 https://github.com/azerothcore/azerothcore-wotlk/issues/7364

### AC rev. hash/commit

https://github.com/chromiecraft/azerothcore-wotlk/commit/420ca2e2b57216b90f93e2deb870946b420a72df

### Operating system

Ubuntu 20.04

### Modules

- [mod-ah-bot](https://github.com/azerothcore/mod-ah-bot)
- [mod-cfbg](https://github.com/azerothcore/mod-cfbg)
- [mod-chat-transmitter](https://github.com/azerothcore/mod-chat-transmitter)
- [mod-chromie-xp](https://github.com/azerothcore/mod-chromie-xp)
- [mod-desertion-warnings](https://github.com/azerothcore/mod-desertion-warnings)
- [mod-duel-reset](https://github.com/azerothcore/mod-duel-reset)
- [mod-eluna-lua-engine](https://github.com/azerothcore/mod-eluna-lua-engine)
- [mod-ip-tracker](https://github.com/azerothcore/mod-ip-tracker)
- [mod-low-level-arena](https://github.com/azerothcore/mod-low-level-arena)
- [mod-multi-client-check](https://github.com/azerothcore/mod-multi-client-check)
- [mod-pvp-titles](https://github.com/azerothcore/mod-pvp-titles)
- [mod-pvpstats-announcer](https://github.com/azerothcore/mod-pvpstats-announcer)
- [mod-queue-list-cache](https://github.com/azerothcore/mod-queue-list-cache)
- [mod-server-auto-shutdown](https://github.com/azerothcore/mod-server-auto-shutdown)
- [lua-carbon-copy](https://github.com/55Honey/Acore_CarbonCopy)
- [lua-custom-worldboss](https://github.com/55Honey/Acore_CustomWorldboss)
- [lua-level-up-reward](https://github.com/55Honey/Acore_LevelUpReward)
- [lua-recruit-a-friend](https://github.com/55Honey/Acore_RecruitAFriend)
- [lua-send-and-bind](https://github.com/55Honey/Acore_SendAndBind)
- [lua-temp-announcements](https://github.com/55Honey/Acore_TempAnnouncements)
- [lua-zonecheck](https://github.com/55Honey/acore_Zonecheck)
- [lua-zone-debuff](https://github.com/55Honey/Acore_ZoneDebuff)

### Customizations

None

### Server

ChromieCraft

### PR DESCRIPTION
**Pull request by UltraNix**
The mobs should attack the pike if there is no players around
The boss should not despawning after summon
Urok's Tribute Pile with Pike and Head should disapear after Urok appear
Fixes #9023

## Issues Addressed:

- Closes #9023

## Tests Performed:

- Tested ingame.

## How to Test the Changes:

More info [here](https://github.com/azerothcore/azerothcore-wotlk/issues/9023)

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by UltraNix** [comment]
@Shineslippers @patou01 Please test it.

File: src/server/game/Entities/Object/Object.cpp [hunk 1]
```diff
@@ -2566,7 +2566,8 @@ void WorldObject::GetContactPoint(const WorldObject* obj, float& x, float& y, fl
     // angle to face `obj` to `this` using distance includes size of `obj`
     GetNearPoint(obj, x, y, z, obj->GetObjectSize(), distance2d, GetAngle(obj));
 
-    if (fabs(this->GetPositionZ() - z) > 3.0f || !IsWithinLOS(x, y, z))
+    // Exclude gameobjects from LoS calculations
+    if (fabs(this->GetPositionZ() - z) > 3.0f || (GetTypeId() != TYPEID_GAMEOBJECT && !IsWithinLOS(x, y, z)))
```
Comment (Nyeriah) [inline src/server/game/Entities/Object/Object.cpp:2570]: I'm sorry if it's a dumb question, but this won't affect Sapphiron/Sindragosa's tombs, right?

Comment (UltraNix) [reply src/server/game/Entities/Object/Object.cpp:2570]: It won't.

**Comment by Shineslippers** [comment]
Never tested nothing before, but if instruction is good i have merge conflict and have to ask you to solve it

**Comment by temperrr** [comment]
Tested:

- [x] The mobs should attack the pike if there is no players around
- [x] The boss should not despawning after summon
- [ ] Urok's Tribute Pile with Pike and Head should disapear after Urok appear -  not happening , may be because I broke event

I walked up the ramp to see if monster attacked the pike, next waves one of the back spawns would always go up the ramp, not towards me or the pike. No clue where they went, Urok did spawn.

image

**Comment by UltraNix** [comment]
Just test normally.

**Comment by patou01** [comment]
I'll go through the event tonight to see how it goes.

**Comment by patou01** [comment]
- [ ] Attacking a mob puts the others in combat (so the "CallForHelp()" or whatever is not ignored, it seemed it was only on enforcers)
- [ ] Pathing is absolutely fucked, I doubt that's your mistake though
- [ ] Mobs don't go to the pike after a reset, they just stand there swinging their weapon. It's ugly and could be seen by the players in case of wipe (while waiting to ankh for instance).
- [ ] The timer needs to be decided on how long they can swing at the pike

**Comment by UltraNix** [comment]
`Attacking a mob puts the others in combat (so the "CallForHelp()" or whatever is not ignored, it seemed it was only on enforcers)` It's normal in dungeons.

`Pathing is absolutely fucked, I doubt that's your mistake though` It's not related to it at all.

**Comment by patou01** [comment]
> It's normal in dungeons.

While not certain of this specific event, it used to be a strat to let the mobs attack the shields (eg BM, or the one from northrend that's the same as BM), so I'm not sure that's "normal", but maybe acceptable here.

Count of swings doesn't need to be exact, but we need some kind of okay estimation. It seemed a lot here.

**Comment by UltraNix** [comment]
@temperrr `Urok's Tribute Pile with Pike and Head should disapear after Urok appear - not happening , may be because I broke event`
It's the same as @patou01 - broken pathfinding here. I don't know what is the problem, but it's not covered by this PR.

**Comment by temperrr** [comment]
Up to you what to do with this then Ultranix

**Comment by UltraNix** [comment]
@temperrr @patou01 Problems with "escaping to nowhere" mobs and swing animation after evade should be fixed.
The whole script is ready.

**Comment by patou01** [comment]
Just tried again.

- [ ] tribute pile takes a while to respawn, not sure if it would be better to respawn on fail?
- [ ] mobs run back to their spawn point before going to the pike on resetting (minor)
- [ ] if you wipe after Urok has spawned and there's still mobs present, they start acting funny attacking nothing. Maybe despawn them on reset if the event is failed?

Rest seems fine, it's quite obvious that it's a private server but "eh". I don't think we can do it much better.

Video here, might take a few minutes to be ready to be viewed. https://drive.google.com/file/d/1W1L5eVAd-unBBhCSMjh3i_3GuNFGYy8t/view?usp=sharing

**Comment by UltraNix** [comment]
@patou
`tribute pile takes a while to respawn, not sure if it would be better to respawn on fail?`
It's okay - https://github.com/azerothcore/azerothcore-wotlk/issues/9023#issuecomment-963494153

`mobs run back to their spawn point before going to the pike on resetting (minor)`
Also okay - it is shown in the video https://youtu.be/L8a-9CYQk1g?t=4165

**Comment by Nyeriah** [comment]
Those SAI changes will need to be documented (wiki needs an update)

**Review by Nyeriah — APPROVED** [review]

**Pull request merged**